src/shims/posix/fs.rs

```diff
@@ -1280,30 +1277,38 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     immty_from_uint_checked(0u128, c_ushort_layout)?, // d_reclen
                     immty_from_int_checked(file_type, c_uchar_layout)?, // d_type
                 ];
+                let mut name = dir_entry.file_name(); // not a Path as there are no separators!
+                name.push("\0"); // Add a NUL terminator
+                let name_bytes = os_str_to_bytes(&name)?;
+                let name_offset = imms.iter().map(|imm| imm.layout.size.bytes()).sum::<u64>();
+                let size =
+                    name_offset.checked_add(u64::try_from(name_bytes.len()).unwrap()).unwrap();
+
+                let entry = this.malloc(size, /*zero_init:*/ false, MiriMemoryKind::C)?;
```

Comment:
`MiriMemoryKind::Machine` seems like a more suitable memory kind? That will also ensure that if the user calls `free`, we have UB (even if they never call `readdir` or `close` on that FB again).

Reply:
Oh, but `Machine` is allowed to leak. We don't want that.

So maybe a new memory kind is in order then, or `Env` should be renamed to something that also encompasses this.

Reply:
Or maybe just `Runtime`?